library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Sum.hs

```diff
@@ -0,0 +1,224 @@
+        address' = sumPtr,
+        indincies' = Block.constant32List [0, 0]
+      }
+
+getVariantPtr :: Types.Define m => AST.Operand -> m AST.Operand
+getVariantPtr sumPtr =
+  Block.getElementPtr
+    Types.Minimal
+      { type' = Types.pointerOf variantPtrType,
+        address' = sumPtr,
+        indincies' = Block.constant32List [0, 1]
+      }
+
+getSumDesc :: Types.LookupType m => PassTypes.SumName -> m Types.SumDesc
+getSumDesc sumName = do
+  sumTable <- get @"sumTab"
+  case Map.lookup (toSymbol sumName) sumTable of
+    Just sumDesc -> pure sumDesc
+    Nothing ->
+      throw @"err" $
+        Types.NonExistentSumType "typechecker allowed sum of non-existent type"
+
+lookupType :: Types.LookupType m => PassTypes.SumName -> m Type.Type
+lookupType sumName = do
+  getSumDesc sumName >>= pure . fst
+
+oneArgFunctionType :: Type.Type -> Type.Type -> Type.Type
+oneArgFunctionType arg result =
+  Types.pointerOf
+    Type.FunctionType
+      { Type.resultType = result,
+        Type.argumentTypes = [arg],
+        Type.isVarArg = False
+      }
+
+makeCase ::
+  Types.Define m =>
+  PassTypes.SumName ->
+  Type.Type ->
+  [Type.Type] ->
+  AST.Operand ->
+  [AST.Operand] ->
+  [AST.Operand] ->
+  m AST.Operand
+makeCase sumName outputType caseTypes term cases environments = do
+  (sumType, variantDescs) <- getSumDesc sumName
+  let values = map tagConstant [0 .. length variantDescs - 1]
+  let variantNames = map fst variantDescs
+  let variantTypes = map snd variantDescs
+  let expectedCaseTypes = map (`oneArgFunctionType` outputType) variantTypes
+  if caseTypes /= expectedCaseTypes
+    then
+      throw @"err" $
+        Types.MismatchedCaseTypes $
+          "typechecker allowed mismatched case types: expected "
+            <> show expectedCaseTypes
+            <> "; got "
+            <> show caseTypes
+            <> "; outputType "
+            <> show outputType
+            <> "; variantTypes "
+            <> show variantTypes
+    else pure ()
+  indexPtr <- getIndexPtr term
+  index <- Block.load indexType indexPtr
+  variantPtrLoc <- getVariantPtr term
+  variantPtr <- Block.load variantPtrType variantPtrLoc
+  let appliedCases =
+        map
+          ( \(ty, (caseFunc, environment)) -> do
+              castedPtr <- Block.bitCast variantPtr (Types.pointerOf ty)
+              variant <- Block.load ty castedPtr
+              Block.call outputType caseFunc [(environment, []), (variant, [])]
+          )
+          (zip variantTypes $ zip cases environments)
+  Block.generateSwitch outputType index variantNames values appliedCases
+
+-- | Given the name of a sum type and a compiled variant term,
+-- | allocate a sum type and store the given term in it.
+-- |
+-- | WARNING:  The memory allocations herein leak memory, pending our
+-- | implementing garbage collection.
+makeSum ::
+  Types.Define m =>
+  PassTypes.SumName ->
+  PassTypes.VariantName ->
+  Type.Type ->
+  Type.Type ->
+  AST.Operand ->
+  m AST.Operand
```

Comment:
We really need to make data types over these functions, it feels like we pass a lot of arguments to a lot of functions